### ISSUE TITLE
In the estimator of Tensorflow, how does it work when model_fn is called multiple times?

### ISSUE THREAD

**Issue by hanfeisun** [tensorflow/tensorflow#13895]
def model_fn(features, labels, mode, params):
      """Model function for Estimator."""
    
      # Connect the first hidden layer to input layer
      # (features["x"]) with relu activation
      first_hidden_layer = tf.layers.dense(features["x"], 10, activation=tf.nn.relu)
    
      # Connect the second hidden layer to first hidden layer with relu
      second_hidden_layer = tf.layers.dense(
          first_hidden_layer, 10, activation=tf.nn.relu)
    
      # Connect the output layer to second hidden layer (no activation fn)
      output_layer = tf.layers.dense(second_hidden_layer, 1)
    
      # Reshape output layer to 1-dim Tensor to return predictions
      predictions = tf.reshape(output_layer, [-1])
    
      # Provide an estimator spec for `ModeKeys.PREDICT`.
      if mode == tf.estimator.ModeKeys.PREDICT:
        return tf.estimator.EstimatorSpec(
            mode=mode,
            predictions={"ages": predictions})
    
      # Calculate loss using mean squared error
      loss = tf.losses.mean_squared_error(labels, predictions)
    
      # Calculate root mean squared error as additional eval metric
      eval_metric_ops = {
          "rmse": tf.metrics.root_mean_squared_error(
              tf.cast(labels, tf.float64), predictions)
      }
    
      optimizer = tf.train.GradientDescentOptimizer(
          learning_rate=params["learning_rate"])
      train_op = optimizer.minimize(
          loss=loss, global_step=tf.train.get_global_step())
    
      # Provide an estimator spec for `ModeKeys.EVAL` and `ModeKeys.TRAIN` modes.
      return tf.estimator.EstimatorSpec(
          mode=mode,
          loss=loss,
          train_op=train_op,
          eval_metric_ops=eval_metric_ops)

Above is an example of the model_fn used by Tensorflow's [Estimator][1].

As mentioned in the tutorial, this model_fn could be called in different context (train, predict, evaluate). However, I'm a bit confused, because each time the model_fn is called, **instead of reusing existing graph, it seems to create a new graph.(or create new node in the graph)**

For example, firstly I called model_fn under TRAIN mode, then I called model_fn with PREDICT mode. How can I make sure the PREDICT one is reusing the weight of the trained values?

  [1]: https://www.tensorflow.org/extend/estimators

**Comment by hanfeisun** [comment]
Is it true that the code in the tutorial forgets to add "reuse=True" as the parameter of layer functions?

**Comment by drpngx** [comment]
Right, that might be a problem.


@MarkDaoust WDYT?

**Comment by MarkDaoust** [comment]
The code and doc is correct, it just doesn't emphasize enough that it rebuilds the graph from scratch for each method, and for evaluate/predict is loads variables from the most recent checkpoint.

Generally we're moving away from the whole `name_scope` system for reuse, the future of variable sharing is layer objects, if you want to reuse a variable reuse the layer object.

We're rewriting this doc right now, I'll make sure it's more clear about this.

**Comment by drpngx** [comment]
Closing since the doc will be obsolete soon. Please check the new doc once it's out. Thank you both!

**Comment by FrancescoSaverioZuppichini** [comment]
Any idea when the new doc will be out? I have still the problem, the estimator is virtually useless if it creates the same graphs every time I need to evaluate something. 

Cheers,

Francesco Saverio

**Comment by MarkDaoust** [comment]
@FrancescoSaverioZuppichini: The updated docs will be available in 1.5, but drafts are already in the repo, [this doc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/docs_src/get_started/checkpoints.md
) describes how checkpointing works with estimators.

In the mean time, let me summarize:

@hanfeisun: Yes, each (`train`, `evaluate`, `predict`) method rebuilds the `graph`. But the estimator **always** loads the checkpoints from the `model_dir` into the graph before doing anything. In fact `evaluate` and `predict` can't be used until after you've called `train` at least once, so that there is a checkpoint to load. 


The `reuse` layer argument does something different. It's an outdated way to use the same layer on two different inputs. A much better way to have this effect is to use the layer classes, and just reuse the layer object. I believe the following two code blocks are equivalent:

```
out0 = tf.layers.dense(x, 10)
out1= tf.layers.dense(y,10, reuse=True)
```

```
my_layer = tf.layers.Dense(10)
out0 = my_layer(x)
out1 = my_layer(y)
```

I hope this helps, but in the future this sort of question is probably better suited for stack overflow.

**Comment by FrancescoSaverioZuppichini** [comment]
So I am forced to used the layer library? Is there any way to set a flag inside the model_fn? Or to have access to the estimator graph and directly used a scope do build in the same graph? Also, I am wondering how to run validation, assuming I want to train for one epoch and then run the model on the validation set I have to first call `.train` and then `.evaluate` that will re-build the graph slowing everything. Any clever way to do so?

**Comment by MarkDaoust** [comment]
> So I am forced to use the layer library?

No, we encourage you to use `layers` to wrap trainable state, but whether or not you use `layers` has no effect on the checkpoint reloading logic.

> I have to first call `.train` and then `.evaluate` that will re-build the graph slowing everything.

If you want to run repeated evaluations while training, consider the [train_and_evaluate](https://www.tensorflow.org/api_docs/python/tf/estimator/train_and_evaluate) function. 

If you're working outside of estimators, [`Datasets`](https://www.tensorflow.org/programmers_guide/datasets) do have advanced iterator types that allow you to swap input pipelines without rebuilding the graph. but ofter the changes you want to make to the graph between `train`, `evaluate`, and `predict` are more extensive than just the input stream.

**Comment by FrancescoSaverioZuppichini** [comment]
Thank you, really. Definitely, the API could have been better implemented. I am used to the Dataset, I could wrap the `Dataset` inside the input function but I am not sure how to swap between two datasets using the `make_initializable_iterator` since `train_and_evaluate` have not any hooks. 

One last question, it is not clear to me how to avoid the re-creation of the graph using a generic implementation (assuming no `layers` library). Should I always set `reuse=True`, is there a faster way? Maybe with a scope? 

I am sorry to bother you again, but really, the doc is so confusing that I could not find any answers to my questions.

Thank you again :)

**Comment by FrancescoSaverioZuppichini** [comment]
Hello again,

Is there any news? I saw there is a pre-release of the 1.5 version but I was not able to find how to avoid the model re-creation at every call of train/evaluate.

Cheers

**Comment by MarkDaoust** [comment]
Well that's kind of my point. If you're using estimators you have to rebuild it on each call to `train`, `evaluate` or `predict`.

The only way around that is to use a different model abstraction, or roll your own.

For example [the Eager examples](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/eager/python/examples) may provide the flexibility you're looking for.

**Comment by FrancescoSaverioZuppichini** [comment]
As far as I understood they said they will fix it in 1.5

**Comment by FrancescoSaverioZuppichini** [comment]
Dear @MarkDaoust the eager execution cannot be used with the Estimator. Also ` train_and_evaluate` will run evaluation AFTER training. I want to have a validation test that must be run at every train step but this seems impossible to do with the Estimator, maybe a better design could have solved the problem. Any ideas? I want to run one train step and one evaluation step WITHOUT loading the graph in order to also plot the validation score

**Comment by tengerye** [comment]
I have spent a week on my project based on Estimator until I stopped here. I am an early developer and had full faith in it. After having encountered this countless times, my heart is broken. If we can't evaluate during training, then **why is Estimator useful**?

**Comment by FrancescoSaverioZuppichini** [comment]
I also stop to use it. It is a total crap.

**Comment by MarkDaoust** [comment]
>  If we can't evaluate during training, then why is Estimator useful?

Other possible approaches:

 * Alternate `train` and `eval` calls
 * In a separate process, monitor the `model_dir` for checkpoints and run `eval` on each time a new one appears.

**Comment by FrancescoSaverioZuppichini** [comment]
That is more a hack than a solution

**Comment by drpngx** [comment]
CC @martinwicke

**Comment by martinwicke** [comment]
@ispirmustafa FYI also.

Sorry you all are frustrated. It also appears the documentation is adding to your confusion. Let me clarify a few points:

* Estimator will evaluate during training. See the train_and_evaluate documentation. It sounds like you're doing single machine execution.
* You can simply pass a function returning a dataset to Estimator. There shouldn't be a problem swapping out inputs.
* Estimator will run the `model_fn` every time you call `train`, `predict`, or `evaluate`. 
* It will automatically reload the last checkpoint (or you can specify one)
* You don't have to worry about reuse between train and eval. As long as your variables are called the same (most likely, simply by using the same or similar enough code to create the layers or raw variables), they will be automatically shared between training and evaluation.
* Personally, I would stay away from `reuse`, and I would never use `variable_scope`. If you must (or have an existing model using this), that's fine. You can use either of these features to reuse variables within your model. Note that has absolutely no effect on whether variables are shared between training and evaluation. Unless you take steps to prevent it, variables are always shared between training and evaluation.

**Comment by ispirmustafa** [comment]
an addition to @martinwicke, `Estimator` creates a new graph for every call of `model_fn`. That makes variable names kept same for each `mode`.

**Comment by tengerye** [comment]
I am kind of confuse again. If `Estimator` creates a new graph for every call of `model_fn`, then each of `train` and `evaluate` has a graph. A loop iterates `train` and `evaluate` will create two new graphs every time? @ispirmustafa @martinwicke

**Comment by FrancescoSaverioZuppichini** [comment]
@tengerye  yes! That's the problem

**Comment by martinwicke** [comment]
Yes, if you call train and evaluate in a loop, you will create two new graphs every time. You will also read and write a checkpoint in each iteration of your loop. If you epoch is long, you probably want a checkpoint every time anyway so that's not a big deal, but if you train on less than an epoch in each iteration, such loops are a bad idea.

@FrancescoSaverioZuppichini I'm afraid I lost track of what the actual problem is.

**Comment by tengerye** [comment]
Thank you sincerely for your kind answer @martinwicke. According to your reply, `train_and_evaluate` may suit my requirement. I believe @FrancescoSaverioZuppichini and me are expecting more flexible implementation that we could write some snippets of codes to replace a component instead of re-writing everything. 
Thank you again.

**Comment by martinwicke** [comment]
@tengerye which component would you like to replace?

Many of the restrictions of Estimator are in place so you can scale up without having to change your code. That makes local training less convenient, but it avoids a lot of pitfalls later.

**Comment by tengerye** [comment]
@martinwicke I see your point now. I thought it would be better if training and evaluation could progress interactively under single machine execution. I think maybe it it better to provide an independent implementation especially for single machine? Since a lot of people work on single machine at the beginning.

**Comment by martinwicke** [comment]
For single machine training you can start with tf.keras.Model. Once your ready to scale, use that Model in a model_fn, or use model_to_estimator. 

Maybe there's a way to do interactive training in a way that still allows you to scale without forcing people to rethink their models, but I have no solution to that (the workflow above using Keras is the best I've seen).

**Comment by datlife** [comment]
@martinwicke May you elaborate the argument that _"... but if you train on less than an epoch in each iteration, **such loops are a bad idea."**_ .  In many machine learning methods, shouldn't we need to evaluate the model after every epoch? 

AFAIK, Keras has the idea to evaluate model every epoch  in its `fit_generator` and `fit`. Particularly, one use case (that been showed in TensorFlow official code [here](https://github.com/tensorflow/models/blob/ce4459762727264f35a00a7ff1f8151380ea681d/official/resnet/resnet.py#L605) . Here is my example using K-fold cross validation for `Estimator`. : 

```python
for _ in range(training_epochs // epochs_per_eval):
  # #####################
  # K-fold cross validation
  # #####################
  train_features, train_labels, eval_features, eval_labels = data_provider.split_training_data(
      test_size=0.2,
      shuffle=True)

  # The pipeline below is similar to `fit_generator` in Keras
  # ##############
  # Training cycle
  # ##############
  estimator.train(
      input_fn=lambda: data_provider.get_input_fn(
          mode=tf.estimator.ModeKeys.TRAIN,
          features=train_features,
          labels=train_labels,
          batch_size=batch_size,
          parse_fn=parse_sample_fn,
          preprocess_fn=preprocess,
          steps_per_epoch=steps_per_epoch * epochs_per_eval,
          shuffle_buffer=shuffle_buffer,
          num_parallel_calls=cpu_cores),
      steps=steps_per_epoch * epochs_per_eval,)

  # ################
  # Evaluation cycle
  # ################
  eval_result = estimator.evaluate(
      input_fn=lambda: data_provider.get_input_fn(
          mode=tf.estimator.ModeKeys.EVAL,
          features=eval_features,
          labels=eval_labels,
          batch_size=batch_size,
          parse_fn=parse_sample_fn,
          preprocess_fn=preprocess,
          steps_per_epoch=200,
          shuffle_buffer=None,
          num_parallel_calls=cpu_cores),
      steps=200)
  print(eval_result)
print("---- Training Completed ----")
```

One of the issues ( I think) with `Estimator` in this approach is : 
* If `epochs_per_eval` and `steps_per_epoch` is relative small, the overhead to create the Graph for training and evaluation in each iteration in a loop is huge.

My question is:
* Is it possible to share the Graph in [`_train_model`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/estimator/estimator.py#L817) and [`_evaluate_model`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/estimator/estimator.py#L910), instead of creating them everytime ? Thanks for reading.

**Comment by martinwicke** [comment]
> My question is:
>
> - Is it possible to share the Graph in _train_model and _evaluate_model, instead of creating them 
>   everytime ? Thanks for reading.

The reason for this strict separation is that we want to guarantee that the model can be scaled up and trained in a distributed fashion. Therefore we make sure that the behavior of an `Estimator` is consistent, whether you train it in local mode or on a cluster of machines. If we allowed you to share the graph between training and eval, it would be very easy to introduce subtle mistakes that would break the training once you run it distributed, so we do not allow it.

No. I understand that this makes life harder in "local" mode. If you are not interested in distributed training, I recommend using `tf.keras.Model`. You can always go back to an Estimator using `model_to_estimator` if you change your mind.

**Comment by FrancescoSaverioZuppichini** [comment]
@martinwicke Thank you for your answer but I think you just did not think about that. I cannot believe it is too hard to add a parameter, something like `reset=False` to the Estimator.

**Comment by martinwicke** [comment]
It would be hard to do in `Estimator`, but you could probably do it in `train_and_evaluate`, changing how it evaluates in local mode. We had actually discussed this, and it would be feasible for `train_and_evaluate` to build both train and eval graphs against the same set of variables, not load checkpoints, and switch between training and evaluation in memory. This would still not share the graphs (which wouldn't be a good idea given the distribution concerns), but it would avoid reloading. 

We never prioritized this since it's unclear to me how much you gain from this. For small models, checkpoint loading isn't a huge cost, and for large models, checkpoint loading again isn't a huge cost compared to the cost of training.

**Comment by FrancescoSaverioZuppichini** [comment]
I can assure you that it is impossible to do cross-validation since loading the model takes too long. Probably the question is still unclear, all these people are asking to train the model for one epoch, run the validation set, and restart until the train loop is finished. 

So assuming we have 100 epochs, we need to run 100 times the estimator in train mode and 100 times the estimator in evaluate mode. This adds up to 200 times the model is restored. Come on guy! You are Google, are you? `train_and_evaluate` will just evaluate after the train is done. Meaning, that it will first train for, following our example, 100 epochs and then run the evaluation one once! 

I find really hard to believe that, after all these people explain the same problem, is still so unclean. Without any way to properly do cross-validation, the Estimator is almost a useless piece of software since we can not even see if our model is overfitting or not using a validation set.

@martinwicke Thank you to quickly reply to us, I hope now the problem is cleaner.

**Comment by martinwicke** [comment]
`train_and_evaluate` *will* evaluate after each epoch (or more precisely in intervals you can set yourself). As you say it will also load the checkpoint after each epoch. So I do indeed not understand the problem. From your description I think it would solve your problem just fine, unless there is a performance problem, and I have tried to explain why I don't think performance is an issue here. 

As I explained above, Estimator does make some tradeoffs where it enforces things that are unnecessary in order to ensure that the transition to distributed training will work smoothly. If Estimator is not suitable for your use case (and especially if you do not need distributed training), you should look into tf.keras.Model.

I do suggest you keep your comments civil, otherwise it will be hard to have a constructive discussion.

**Comment by FrancescoSaverioZuppichini** [comment]
@martinwicke I am sorry if something I wrote seems uncivil to you. The doc about `train_and_evaluate` does not says anything about a parameter to set the intervals https://www.tensorflow.org/api_docs/python/tf/estimator/train_and_evaluate 

I was trying to use Estimator with a model to generate text, even with just one layer of LSTM with 256 cells, it takes forever to load the graph after each epoch.  So for now, I just give up. Paradoxically the problem is even highlighted when using small models to test/play with. 

I agree that the model must be loaded to ensure consistency in a distributed training, but still, I can not see why such an easy and important feature was not implemented.

**Comment by martinwicke** [comment]
You are correct, the information about the local run behavior for evaluation is missing from the `train_and_evaluate` docstring (and also from the EvalSpec docstring; @xiejw FYI, we should add this). 

`train_and_evaluate` will evaluate every `eval_spec.throttle_sec` seconds. It will stop after `train_spec.max_steps`. 

I am still surprised a small model takes long to load (how long is "forever"?). Could it be your Python code? I.e. does it also take long before it starts the first iteration, before there is a checkpoint? It will create the graph using its Python code, not load it from the checkpoint, so that can be a source of slowness.

**Comment by nxphi47** [comment]
I agree with @FrancescoSaverioZuppichini . I read the docstring and the implementation itself, `train_and_evaluate` is just simply a for loop iterating between estimator.train and estimator.evaluate. 

I never try distributed training cos I can't afford it. By doing train and evaluate with estimators is painful even though it does offer many other good features. 

Each restoring from checkpoint and kick start the session takes ridiculously long time. Unless you are training ImageNet with millions images on models of tens of layers which consumes your whole weekend just for one epoch, re-build the graph every time is so time consuming. In my case, it takes about 5-10% training time cos my dataset is small.

What many peoples here want is (I try to summarize again):
for e in range(epochs):
      train for whole training set.
      evaluate on train set, save loss, accuracy, metrics... to tensorboard at ./summary_train/
      evaluate on val set, save loss, accuracy, metrics... to tensorboard at ./summary_val/

At the end of the day, we will have a trained model, a chart indicates the loss and accuracy of both train and val metrics to measures overfitting and so on.

Estimator does this with 2 or 3 times re-building the graph and loading checkpoints.

About graph consistency between training and evaluation, one can simply build a 2-in-1 graph with a `is_training` placeholder indicating when should the graph execute in which mode. `tf.layers.barch_norm` does allow passing `training` with a bool tensor. Or we can use `tf.cond` to branch between each mode. But estimators do not allow this anyway.

It is OK to keep "training things" like gradients, optimizer parameters in validation mode instead of release them and re-build the graph again.

**Comment by janpfeifer** [comment]
A little bit orthogonal, but I thought it may address one of @FrancescoSaverioZuppichini 's issue of reloading the model for evaluation at each epoch (or whatever frequency you configure):

I often run things locally (if data is not huge), but most of the time I run different processes (both locally) for training and evaluating.

So one would run with "--train" for the 100 (?) epochs, saving checkpoints at every X time/epochs.

And the evaluation process would monitor for new checkpoints, load and evaluate them in parallel, without stopping the training. This can be done with [tf.contrib.Experiment](https://www.tensorflow.org/api_docs/python/tf/contrib/learn/Experiment) -- see [continuous_eval()](https://www.tensorflow.org/api_docs/python/tf/contrib/learn/Experiment#continuous_eval) -- and [tf.contrib.learn_runner](https://www.tensorflow.org/api_docs/python/tf/contrib/learn/learn_runner/run) -- with `schedule="continuous_eval"` -- with not many extra lines of code.

I hope it helps.

Btw, how big are your checkpoints ? When you say it takes a long time to load, how much is it roughly ? (10s, 1min, 10min, 30min, 1hour?)

Oh, also, how are you doing cross-validation ? Sorry I didn't figure from the code/description ... 

cheers

**Comment by jperl** [comment]
Our model takes up nearly all of our GPU memory so we cannot instantiate two of the graphs on one computer. However we want to periodically evaluate sharing the same variables. I guess we cannot use Estimators for this?

**Comment by datlife** [comment]
@jperl: you can, however, expect the tf.Graph will be reloaded from GPU every time you call Estimator.evaluate. 

It can be slow sometime, especially when your steps_per_eval is small.

**Comment by ldscr** [comment]
so the only way avoid reloading every time is to use the keras.Model?

**Comment by fgervais** [comment]
I'd like to get a bit more information on this problem but from the predict() point of view.

I posted on stackoverflow but really there is not much support there for now. I can see why people would try here.
 
https://stackoverflow.com/questions/49966447/keep-the-tensorflow-estimator-in-memory-while-waiting-for-live-prediction-inputs

It look like when running the estimator live and doing one prediction at a time (as data comes in) most of the prediction time is wasted on reloading the model.

In my use case I'd like to do the [prediction, action, prediction, action, ...] as fast as possible.

Is there a way around having the model reloaded in between predictions?

**Comment by ispirmustafa** [comment]
Hi @fgervais ,
for predict could you please check out [predictor.from_estimator](https://www.tensorflow.org/api_docs/python/tf/contrib/predictor/from_estimator)

**Comment by fgervais** [comment]
I'll give it a try thank you

**Comment by formigone** [comment]
To clarify:

> train_and_evaluate will just evaluate after the train is done. Meaning, that it will first train for, following our example, 100 epochs and then run the evaluation one once!

That is incorrect. `train_and_evaluate` trains for a configurable amount of iterations, then runs an evaluation until the specified `input_fn` finished. Then it starts training again by restoring the weights that were used for the eval, and consuming the input stream from where it left off. There might be some minor technicalities in my summarized explanation, but it is conceptually accurate. 

I've been training reasonably models on a single laptop using `train_and_evaluate` since version 1.2, and I think it works extremely well. The time it takes to rebuild the graph and restore parameters is negligible when you take into account all the time that's spent actually training.

**Comment by dage** [comment]
@ispirmustafa @fgervais I'm also giving this a try. It feels as if I'm almost there, but I'm still struggling with a couple of issues. I've just posted a question with a minimalistic example on stackoverflow:  
https://stackoverflow.com/questions/50111636/unable-to-use-core-estimator-with-contrib-predictor

**Comment by fgervais** [comment]
@ispirmustafa the predictor worked, thank you.

It it helps, I added a bit of information on how I did it [here](https://stackoverflow.com/questions/49966447/keep-the-tensorflow-estimator-in-memory-while-waiting-for-live-prediction-inputs/).

**Comment by dage** [comment]
I made a public repo which provides a complete example of using the predictor with a custom core estimator, where it compares prediction results and performance between estimator.predict() and predictor():  

https://github.com/dage/tensorflow-estimator-predictor-example

**Comment by ispirmustafa** [comment]
Hi @dage @fgervais 
Thank you for update. 
If you want to update pydoc of estimator.predict to give your examples I'll be happy to review.

**Comment by rabeehkarimi** [comment]
Hi, I need to see the plots for validation convergence as I train using estimators, and cannot see how to pass validation set to the train function, or how to compute validation losses, I also tried to write my own, without estimators, and always gets the error that graph cannot be larger than 2Gig, could you please help. thanks.

**Comment by martinwicke** [comment]
@rabeehkarimi use tf.estimator.train_and_evaluate.

**Comment by cyinv** [comment]
@formigone I'm afraid that your understanding of how input_fn works is not accurate. 
According to the instruction:
> It is also recommended to train the model a little longer, say multiple epochs, before
  performing evaluation, as the input pipeline starts from scratch for each
  training.

So i guess with estimator api, if I want to evaluate within one epoch without resetting the input flow, I have to somehow store the iterator status inside the input_fn?

**Comment by martinwicke** [comment]
If you're using Estimator in a distributed system, evaluation happens in parallel to training, and not strictly in between epochs. 

In local mode, to evaluate, say, twice per epoch, you can either use a variable in the input_fn, and store state there, or you can store that state in a hook. Either way, this is somewhat complicated. 

Another option (and the one used by most of our production pipelines) is to shuffle the inputs. That way, you will get a different set of inputs every time you run, and you can simply evaluate after a given set of steps and then evaluate, without taking special care of the input state. If you do this, be careful that you don't only shuffle the beginning of the data.

**Comment by shanest** [comment]
What about the following solution, where we pass in the same estimator to a `SessionRunHook` and evaluate it every certain number of steps (which could be calculated to be roughly one epoch).  As far as I can tell, since `train` is only called once, it doesn't recreate its graph [though the calls to `evaluate` will].  Does this break things?

```python
class EvalEarlyStopHook(tf.train.SessionRunHook)

    def __init__(self, estimator, eval_input_fn, num_steps):

        self._estimator = estimator
        self._input_fn = eval_input
        self._num_steps = num_steps

    ...

    def after_run(self, run_context, run_values):

        global_step = run_values.results['global_step']
        if (global_step-1) % self._num_steps == 0:
            ev_results = self._estimator.evaluate(input_fn=self._input_fn)

            if do_stuff_with_results:
                 run_context.request_stop()
```

With the main call being something like:

```python
estimator.train(input_fn=train_input_fn,
                        hooks=[EvalEarlyStopHook(estimator, eval_input_fn, num_steps))
```

**Comment by jperl** [comment]
Just came across this awesome API in TF 1.9. Thank you TF team!!!

https://www.tensorflow.org/versions/r1.9/api_docs/python/tf/contrib/estimator/InMemoryEvaluatorHook

**Comment by ispirmustafa** [comment]
Hi @jperl ,
Please let us know your feedback based on your usage.

**Comment by jperl** [comment]
@ispirmustafa I just upgraded our project to TF 1.9 and the InMemoryEvaluatorHook. It works fantastically :ok_hand: and has spec up the training eval loop significantly.

There is just one minor bug that I made a PR for https://github.com/tensorflow/tensorflow/pull/20818. I will work on adding a unit test later today.

**Comment by cecabert** [comment]
@jperl Correct me if I'm wrong, the `InMemoryEvaluatorHook` solves the issue of validating the *network* while training it, but the `tf.estimator.Estimator` still creates multiple graphs (one for train + one for evaluation) ?

**Comment by martinwicke** [comment]
That is correct.

**Comment by tavramov** [comment]
Hi @martinwicke ,
`InMemoryEvaluatorHook` looks like a great way to evaluate while training but I am running into issue where all eval_metrics are evaluated to 0.0 after the very first time the evaluation hook is triggered. 
Here is a snippet of my output when `every_n_iter=50`:

INFO:tensorflow:Saving dict for global step 0: accuracy = 0.158545, global_step = 0, loss = 3.36605
INFO:tensorflow:Saving dict for global step 50: accuracy = 0.0, global_step = 50, loss = 0.0
INFO:tensorflow:Saving dict for global step 100: accuracy = 0.0, global_step = 100, loss = 0.0
INFO:tensorflow:Saving dict for global step 150: accuracy = 0.0, global_step = 150, loss = 0.0

My suspicion is that it does not work because my  `input_fn` is defined as `tf.estimator.inputs.numpy_input_fn` with `num_epochs=1`.

The exact definition of the input function that I use to create the `InMemoryEvaluatorHook` is:
```
    eval_input_fn = tf.estimator.inputs.numpy_input_fn(
        x={"x": eval_data},
        y=eval_labels,
        num_epochs=1,
        batch_size=25,
        shuffle=False)
```
How can I confirm if my input_fn is really to blame here? Any ideas how to get around this? Thanks

**Comment by martinwicke** [comment]
Yes, I think your hunch is correct. You probably have to use `num_epochs=None` and set the appropriate amount of eval steps to stop evaluation.

**Comment by ispirmustafa** [comment]
Hi @tavramov ,
Most probably it's not related to num_epoch.
Could you please create a simple script to reproduce this issue? Also do you mind to open a new issue?

**Comment by tavramov** [comment]
Issue can be easily reproduced. I just inserted `InMemoryEvaluatorHook` to cnn_mnist.py
Also submitted Issue #[21590](https://github.com/tensorflow/tensorflow/issues/21590)
Thanks

**Comment by JerrikEph** [comment]
@shanest does your proposal work? it seemed more readable than `InMemoryEvaluatorHook`, it would be nice if works.

**Comment by JerrikEph** [comment]
@shanest does your proposal work? it seemed more readable than InMemoryEvaluatorHook, it would be nice if works. It might will create another "evaluate" session inside "train" session, I am not sure how it will behave, could you please report its behavior?

**Comment by JerrikEph** [comment]
Although `InMemoryEvaluatorHook ` is not loading checkpoint from disk, [it fetches all variable from "train" graph and feed it to the "evaluation" graph](https://github.com/tensorflow/tensorflow/blob/25c197e02393bd44f50079945409009dd4d434f8/tensorflow/contrib/estimator/python/estimator/hooks.py#L170). It still recreates an evaluation graph and all the variables with it . So the problem of graph recreation remains and the slow loading ckpt from disk is simply replaced with a memory copy. Are there any follow ups on graph reusing in `tf.estimator` ?

**Comment by jperl** [comment]
@JerrikEph If you can I recommend switching to a Keras model / training loop. Then you can create a callback to evaluate using the same model. After switching to Keras models I have not looked back to estimators.

**Comment by satyrswang** [comment]
while do evaluation in the runhook, the eval_metrics calculate each batch and the average the result or calculate after all evaluation dataset? like auc in eval_metrics. Anybody knows this?Please help
